src/gui/entry/EntryAttachmentsWidget.cpp

```diff
@@ -0,0 +1,358 @@
+        if (readOk) {
+            m_entryAttachments->set(fInfo.fileName(), data);
+        } else {
+            errors.append(QString("%1 - %2").arg(fInfo.fileName(), file.errorString()));
+        }
+    }
+
+    if (errorMessage && !errors.isEmpty()) {
+        *errorMessage = tr("Unable to open files:\n%1").arg(errors.join('\n'));
+    }
+
+    return errors.isEmpty();
+}
+
+bool EntryAttachmentsWidget::openAttachment(const QModelIndex& index, QString* errorMessage)
+{
+    const QString filename = m_attachmentsModel->keyByIndex(index);
+    const QByteArray attachmentData = m_entryAttachments->value(filename);
+
+    // tmp file will be removed once the database (or the application) has been closed
+    const QString tmpFileTemplate = QDir::temp().absoluteFilePath(QString("XXXXXX.").append(filename));
+    QTemporaryFile* tmpFile = new QTemporaryFile(tmpFileTemplate, this);
+
+    const bool saveOk = tmpFile->open()
+                        && tmpFile->write(attachmentData) == attachmentData.size()
+                        && tmpFile->flush();
+    if (!saveOk) {
+        if (errorMessage) {
+            *errorMessage = QString("%1 - %2").arg(filename, tmpFile->errorString());
+        }
+        delete tmpFile;
+        return false;
+    }
+
+    tmpFile->close();
+    const bool openOk = QDesktopServices::openUrl(QUrl::fromLocalFile(tmpFile->fileName()));
+    if (!openOk) {
+        if (errorMessage) {
+            *errorMessage = QString("Can't open file \"%1\"").arg(filename);
+        }
+        delete tmpFile;
+        return false;
+    }
+
+    return true;
+}
+
+bool EntryAttachmentsWidget::eventFilter(QObject* watched, QEvent* e)
+{
+    if (watched == m_ui->attachmentsView->viewport() && !isReadOnly()) {
+        switch (e->type()) {
+        case QEvent::DragEnter:
+        case QEvent::DragMove: {
+            QDropEvent* dropEv = static_cast<QDropEvent*>(e);
+            const QMimeData* mimeData = dropEv->mimeData();
+            if (mimeData->hasUrls()) {
+                dropEv->acceptProposedAction();
+                return true;
+            }
+            break;
+        }
+        case QEvent::Drop: {
+            QDropEvent* dropEv = static_cast<QDropEvent*>(e);
+            const QMimeData* mimeData = dropEv->mimeData();
+            if (mimeData->hasUrls()) {
+                dropEv->acceptProposedAction();
+                QStringList filenames;
+                for (const QUrl url: mimeData->urls()) {
+                    QFileInfo fInfo(url.toLocalFile());
+                    if (fInfo.isFile()) {
+                        filenames.append(fInfo.absoluteFilePath());
+                    }
+                }
+
+                QString errorMessage;
+                if (!insertAttachments(filenames, &errorMessage)) {
+                    errorOccurred(errorMessage);
+                }
+
+                return true;
+            }
+        }
+        default:
+            break;
```

Comment:
Statement does nothing, remove default clause.

Reply:
if I delete **default** statement then the compiler will be very dissatisfied)

![image](https://user-images.githubusercontent.com/3104366/34327783-30846ff6-e8d6-11e7-92c9-97e423687c95.png)

Reply:
Ouch. Maybe use an if/else if construct? You are only handling three cases of many and this empty default clause is extremely ugly.

Reply:
Done